### PR TITLE
feat: infer project from explicit service account creds

### LINKED ISSUE #27: Infer project from credentials if not passed explicitly
Split off from #17

See also: https://github.com/googleapis/python-datastore/issues/74
See also: https://github.com/googleapis/python-storage/issues/177

In a "blank slate" design, I would argue that the credentials' project should supersede the one present in the environment.  However, for backward compatibility, we need to use the credentials' project only if the environment variable is not set.   So the order should follow the pattern from `google.auth.default` when inferring project ID:

- Explicit project ID passed to client constructor
- Project ID set in `GOOGLE_CLOUD_PROJECT` envvar
- Project ID from credentials, if present
- Project ID from SDK
- Project ID from GAE
- Project ID from GCE

### PR DESCRIPTION
**Pull request by tseaver**
Closes: #27

See: https://github.com/googleapis/python-storage/issues/177
See: https://github.com/googleapis/python-datastore/issues/74

### CODE REVIEW

**Review by crwilcox — APPROVED** [review]

**Comment by crwilcox** [comment]
@busunkim96 wdyt?

File: google/cloud/client.py [hunk 1]
```diff
@@ -188,26 +190,50 @@ class _ClientProjectMixin(object):
...
+            )
+
+        # Project set on explicit credentials overrides discovery from
+        # SDK / GAE / GCE.
+        if project is None and credentials is not None:
+            project = getattr(credentials, "project_id", None)
+
+        if project is None:
+            project = self._determine_default(project)
+
```
Comment (busunkim96) [inline google/cloud/client.py:224]: I think `google.auth.default()` already treats the environment project id with highest priority. In that case would it be sufficient to just have this bit of code?

https://github.com/googleapis/google-auth-library-python/blob/37141e4dffc2ba3f3f57c5914544fb8b9cf7d017/google/auth/_default.py#L296-L298

Comment (tseaver) [reply google/cloud/client.py:224]: @busunkim96 The intent of duplicating the environment check above is to ensure that a project ID set in the environment overrides the one from the credentials (for backward compatibility).

Only if neither one is set to we kick down to the "platform checks" in `_determine_default` (gcloud SDK, GAE, GCE, etc.), which just calls into `google.auth.default`.

Comment (busunkim96) [reply google/cloud/client.py:224]: Ah I see. This seems fine for now. It looks like we unfortunately don't have a spec on how project ID detection is supposed to work in the libraries. 😢

**Pull request merged**